CorsixTH/Lua/dialogs/resizables/file_browser.lua

```diff
@@ -189,31 +189,47 @@ function UIFileBrowser:UIFileBrowser(ui, mode, title, vertical_size, root, show_
   self.on_top = mode == "menu"
   self.esc_closes = true
   self.resizable = false
+  self.origin = origin
   self:setDefaultPosition(0.5, 0.25)
 
   self:addBevelPanel((h_size - 190) / 2, 10, 190, 20, col_caption):setLabel(title)
     .lowered = true
 
   -- Initialize the tree control
-  self.control = FilteredTreeControl(root, 5, 35, h_size - 10, vertical_size, self.col_bg, self.col_scrollbar, true, show_dates)
+  self.control = FilteredTreeControl(root, 5, 35, h_size - 10, vertical_size, self.col_bg, self.col_scrollbar, true, show_dates, origin)
     :setSelectCallback(--[[persistable:file_browser_select_callback]] function(node)
-      if node.is_valid_file and (lfs.attributes(node.path, "mode") ~= "directory") then
-        self:choiceMade(node.path)
-      end
+      self:checkChoice(node)
     end)
   self:addWindow(self.control)
 
   -- Create the back button.
-  self:addBevelPanel((h_size - 160) / 2, 340, 160, 30, self.col_bg):setLabel(_S.menu_list_window.back)
-    :makeButton(0, 0, 160, 40, nil, self.buttonBack):setTooltip(_S.tooltip.menu_list_window.back)
+  local button_size = 135
+  local indent = math.floor((h_size - (2*button_size))/3)
+  self:addBevelPanel(indent, 340, button_size, 30, self.col_bg):setLabel(_S.menu_list_window.back)
+    :makeButton(0, 0, button_size, 40, nil, self.buttonBack):setTooltip(_S.tooltip.menu_list_window.back)
+
+  self:addBevelPanel(h_size - button_size - indent, 340, button_size, 30,
+  self.col_bg):setLabel(_S.menu_list_window.ok)
+    :makeButton(0, 0, button_size, 40, nil, (--[[persistable:filebrowser_ok_callback]] function()
+      if (self.origin.new_savegame_textbox and self.origin.new_savegame_textbox.text ~= "") or (self.origin.new_map_textbox and self.origin.new_map_textbox.text ~= "") then
```

Comment:
Way too much outside knowledge in this line. Pass in a callback instead of origin; and pass the node to the callback.

Reply:
`origin` is actually a derived class of `file_browser`, so a simple method overridden in the derived class is enough.